### PR TITLE
Generate interfaces and implementations for clients

### PR DESCRIPTION
**Pull request by swankjesse**
The Wire plugin was only generating implementations

### CODE REVIEW

File: wire-library/wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt [hunk 1]
```diff
@@ -112,9 +112,16 @@ class KotlinGenerator private constructor(
    * Returns the full name of the class generated for [service]#[rpc]. This returns a name like
    * `RouteGuideClient` or `RouteGuideGetFeatureBlockingServer`.
    */
-  fun generatedServiceName(service: Service, rpc: Rpc? = null): ClassName {
+  fun generatedServiceName(
+    service: Service,
+    rpc: Rpc? = null,
+    isImplementation: Boolean = false
```
Comment (Egorand) [inline wire-library/wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt:118]: Do we care about breaking this class' API?

File: wire-library/wire-test-utils/src/main/java/com/squareup/wire/schema/RepoBuilder.kt [hunk 2]
```diff
@@ -133,11 +133,10 @@ class RepoBuilder {
...
     val service = schema.getService(serviceName)!!
     val rpc = rpcName?.let { service.rpc(rpcName)!! }
-    val packageName = grpcGenerator.generatedServiceName(service).packageName
     return grpcGenerator.generateServiceTypeSpecs(service, rpc)
-        .map { typeSpec ->
-          FileSpec.builder(packageName, "_")
-              .addType(typeSpec)
+        .map { entry ->
+          FileSpec.builder(entry.key.packageName, "_")
+              .addType(entry.value)
```
Comment (Egorand) [inline wire-library/wire-test-utils/src/main/java/com/squareup/wire/schema/RepoBuilder.kt:139]:
```suggestion
        .map { (typeName, typeSpec) ->
          FileSpec.builder(typeName.packageName, "_")
              .addType(typeSpec)
```

File: wire-library/wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt [hunk 3]
```diff
@@ -149,39 +156,42 @@ class KotlinGenerator private constructor(
...
-    val typeSpecs = mutableListOf<TypeSpec>()
-    if (rpcRole == RpcRole.SERVER) {
-      typeSpecs.add(generateService(service, onlyRpc))
-    } else {
-      typeSpecs.add(generateService(service, onlyRpc, generateClientImplementation = false))
-      typeSpecs.add(generateService(service, onlyRpc, generateClientImplementation = true))
+  fun generateServiceTypeSpecs(service: Service, onlyRpc: Rpc? = null): Map<ClassName, TypeSpec> {
+    val result = mutableMapOf<ClassName, TypeSpec>()
+
+    val (interfaceName, interfaceSpec) = generateService(service, onlyRpc, false)
```
Comment (Egorand) [inline wire-library/wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt:162]:
```suggestion
    val (interfaceName, interfaceSpec) = generateService(service, onlyRpc, isImplementation = false)
```

File: wire-library/wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt [hunk 4]
```diff
@@ -149,39 +156,42 @@ class KotlinGenerator private constructor(
...
+    result[interfaceName] = interfaceSpec
+
+    if (rpcRole == RpcRole.CLIENT) {
+      val (implementationName, implementationSpec) = generateService(service, onlyRpc, true)
```
Comment (Egorand) [inline wire-library/wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt:166]:
```suggestion
      val (implementationName, implementationSpec) = generateService(service, onlyRpc, isImplementation = true)
```

File: wire-library/wire-compiler/src/test/java/com/squareup/wire/schema/WireRunTest.kt [hunk 5]
```diff
@@ -90,13 +90,20 @@ class WireRunTest {
     )
     wireRun.execute(fs, logger)
 
-    assertThat(fs.find("generated")).containsExactly(
-        "generated/kt/squareup/routes/RouteClient.kt")
+    println(fs.find("generated"))
```
Comment (oldergod) [inline wire-library/wire-compiler/src/test/java/com/squareup/wire/schema/WireRunTest.kt:93]:
```suggestion
```

File: wire-library/wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt (hunk 1 above)
Comment (swankjesse) [reply wire-library/wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt:118]: Definitely not.

**Pull request merged**